File: app/src/main/java/com/ipd/taxiu/ui/fragment/collect/SocicalContactFragment.java
package com.ipd.taxiu.ui.fragment.collect;

import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.util.Log;
import android.view.View;

import com.ipd.taxiu.R;
import com.ipd.taxiu.adapter.ContactAdapter;
import com.ipd.taxiu.bean.AttentionBean;
import com.ipd.taxiu.bean.BaseResult;
import com.ipd.taxiu.bean.ContactBean;
import com.ipd.taxiu.bean.ContactListBean;
import com.ipd.taxiu.bean.UserBean;
import com.ipd.taxiu.platform.global.GlobalParam;
import com.ipd.taxiu.platform.http.ApiManager;
import com.ipd.taxiu.presenter.MinePresenter;
import com.ipd.taxiu.ui.ListFragment;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

import rx.Observable;
import rx.functions.Func1;

/**
 * Created by Miss on 2018/7/19
 * 社交
 */
public class SocicalContactFragment extends ListFragment<List<AttentionBean>, AttentionBean> implements MinePresenter.IAttentionView {
    private ContactAdapter mAdapter = null;

    private MinePresenter mPresenter;

    public static SocicalContactFragment newInstance(int TYPE) {
        SocicalContactFragment fragment = new SocicalContactFragment();
        Bundle bundle = new Bundle();
        bundle.putInt("TYPE", TYPE);
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    protected void onViewDetach() {
        super.onViewDetach();
        mPresenter.detachView();
        mPresenter = null;
    }

    @Override
    protected void onViewAttach() {
        super.onViewAttach();
        mPresenter = new MinePresenter();
        mPresenter.attachView(getContext(), this);
    }

    @Override
    protected void initView(@Nullable Bundle bundle) {
        super.initView(bundle);
        progress_layout.setEmptyViewRes(R.layout.layout_empty_contact);
    }

    @NotNull
    @Override
    public Observable<List<AttentionBean>> loadListData() {
        Bundle bundle = getArguments();
        final int type = bundle.getInt("TYPE");
        return ApiManager.getService().attentionList(10, GlobalParam.getUserId(), getPage(), type + 1)
                .map(new Func1<BaseResult<List<AttentionBean>>, List<AttentionBean>>() {
                    @Override
                    public List<AttentionBean> call(BaseResult<List<AttentionBean>> listBaseResult) {
                        List<AttentionBean> fans = new ArrayList<>();
                        if (listBaseResult.code == 0) {
                            fans.addAll(listBaseResult.data);
                            for (int i = 0; i < listBaseResult.data.size(); i++) {
                                Log.i("onRefresh", listBaseResult.data.get(i).NICKNAME + "，关注状态：" + listBaseResult.data.get(i).IS_ATTEN);
                            }
                        }
                        return fans;
                    }
                });
    }

    @Override
    public int isNoMoreData(List<AttentionBean> result) {
        if (result == null || result.isEmpty()){
            if (getPage() == getINIT_PAGE()){
                return getEMPTY_DATA();
            }else {
                return getNO_MORE_DATA();
            }
        }
        return getNORMAL();
    }

    @Override
    public void setOrNotifyAdapter() {
        if (mAdapter == null) {
            mAdapter = new ContactAdapter(getContext(), getData(), mPresenter);
            recycler_view.setLayoutManager(new LinearLayoutManager(getContext()));
            recycler_view.setAdapter(mAdapter);
        } else {
            mAdapter.notifyDataSetChanged();
        }
    }

    @Override
    public void addData(boolean isRefresh, List<AttentionBean> result) {
        getData().addAll(result);
    }

    @Override
    public void onSuccess(String msg,int data) {
        toastShow(msg);
        resetData();
        onRefresh();
    }

    @Override
    public void onFail(@NotNull String errMsg) {
        toastShow(errMsg);
    }

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);
        if (isVisibleToUser) {
            resetData();
            onRefresh();
        }
    }
}

File: app/src/main/java/com/ipd/taxiu/ui/fragment/collect/SocicalContactFragment.kt
package com.ipd.taxiu.ui.fragment.collect

import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import com.ipd.taxiu.R
import com.ipd.taxiu.adapter.ContactAdapter
import com.ipd.taxiu.bean.AttentionBean
import com.ipd.taxiu.platform.global.Constant
import com.ipd.taxiu.platform.global.GlobalParam
import com.ipd.taxiu.platform.http.ApiManager
import com.ipd.taxiu.presenter.MinePresenter
import com.ipd.taxiu.ui.ListFragment
import rx.Observable
import java.util.*

/**
 * Created by Miss on 2018/7/19
 * 社交
 */
class SocicalContactFragment : ListFragment<List<AttentionBean>, AttentionBean>(), MinePresenter.IAttentionView {

    companion object {
        fun newInstance(TYPE: Int): SocicalContactFragment {
            val fragment = SocicalContactFragment()
            val bundle = Bundle()
            bundle.putInt("TYPE", TYPE)
            fragment.arguments = bundle
            return fragment
        }
    }


    private var mPresenter: MinePresenter<MinePresenter.IAttentionView>? = null
    override fun onViewAttach() {
        super.onViewAttach()
        mPresenter = MinePresenter()
        mPresenter?.attachView(mActivity, this)
    }


    override fun onViewDetach() {
        super.onViewDetach()
        mPresenter?.detachView()
        mPresenter = null
    }

    override fun initView(bundle: Bundle?) {
        super.initView(bundle)
        progress_layout.setEmptyViewRes(R.layout.layout_empty_contact)
    }

    private val mType by lazy { arguments.getInt("TYPE") }
    override fun loadListData(): Observable<List<AttentionBean>> {
        return ApiManager.getService().attentionList(Constant.PAGE_SIZE, GlobalParam.getUserId(), page, mType + 1)
                .map { listBaseResult ->
                    val fans = ArrayList<AttentionBean>()
                    if (listBaseResult.code == 0) {
                        fans.addAll(listBaseResult.data)
                    }
                    fans
                }
    }

    override fun onResume() {
        super.onResume()
        onRefresh()
    }

    override fun isNoMoreData(result: List<AttentionBean>): Int {
        if (page == INIT_PAGE && (result == null || result.isEmpty())) {
            return EMPTY_DATA
        } else if (result == null || result.isEmpty()) {
            return NO_MORE_DATA
        }
        return NORMAL
    }

    private var clickItemPos = -1
    private var mAdapter: ContactAdapter? = null
    override fun setOrNotifyAdapter() {
        if (mAdapter == null) {
            mAdapter = ContactAdapter(mActivity, data, { position, info ->
                clickItemPos = position
                mPresenter?.attention(info.USER_ID)
            })
            recycler_view.layoutManager = LinearLayoutManager(context)
            recycler_view.adapter = mAdapter
        } else {
            mAdapter?.notifyDataSetChanged()
        }
    }

    override fun addData(isRefresh: Boolean, result: List<AttentionBean>) {
        data?.addAll(result)
    }

    override fun onSuccess(msg: String, isAttent: Int) {
        toastShow(true,msg)
        if (clickItemPos == -1) return
        if (mType == 0){
            data?.removeAt(clickItemPos)
            if (data == null || data!!.isEmpty()){
                progress_layout.showEmpty()
            }
            mAdapter?.notifyItemRemoved(clickItemPos)

        }else{
            data?.get(clickItemPos)?.IS_ATTEN = isAttent
            mAdapter?.notifyItemChanged(clickItemPos)
        }
    }

    override fun onFail(errMsg: String) {
        toastShow(errMsg)
    }
}
